client/src/document/organisms/sidebar/SidebarFilterer.ts

```diff
@@ -0,0 +1,213 @@
+import { splitQuery } from "../../../utils";
+
+export class SidebarFilterer {
+  applyFilter(query: string) {
+    if (query) {
+      return this.showOnlyMatchingItems(query);
+    } else {
+      this.showAllItems();
+      return undefined;
+    }
+  }
+
+  root: HTMLElement;
+  headings: HTMLElement[];
+  parents: HTMLDetailsElement[];
+  links: HTMLAnchorElement[];
+
+  constructor(root: HTMLElement) {
+    this.root = root;
+    this.headings = Array.from(
+      this.root.querySelectorAll<HTMLElement>("li strong")
+    );
+    this.parents = Array.from(
+      this.root.querySelectorAll<HTMLDetailsElement>("details")
+    );
+    this.links = Array.from(
+      this.root.querySelectorAll<HTMLAnchorElement>("a[href]")
+    );
+  }
+
+  showAllItems() {
+    this.links.forEach((link) => this.resetLink(link));
+    this.headings.forEach((heading) => this.resetHeading(heading));
+    this.parents.forEach((parent) => this.resetParent(parent));
+  }
+
+  private resetLink(link: HTMLAnchorElement) {
+    this.resetHighlighting(link);
+    const target = link.closest("li") || link;
+    target.style.display = "inherit";
```

Comment:
As discussed, inline styles have the advantage of having higher specificity, and there is now a `toggleElement()` method that does it to avoid duplication.